aip/client-libraries/4234.md

```diff
@@ -0,0 +1,102 @@
+---
+id: 4234
+state: draft
+created: 2021-02-10
+---
+
+# Common service client mix-ins
+
+Often, APIs can have common features for administering resources that are
+separated into utility services. These common, utility services are shared among
+the discrete product services. These common services each have a
+centrally-defined surface, but individual instances are hosted alongside each
+product service. Thus, the surfaces of each common service hosted by the product
+service are "mixed in" at runtime.
+
+
+## Guidance
+
+Client libraries **may** provide client methods for the common services declared
+to be mixed-in by a product service. Such methods improve the user experience by
+presenting the mix-in methods from a client configured to communicate with the
+product service that hosts the mix-in service.
+
+Should client libraries support mix-in services, the common services that
+**must** be supported include:
+
+- `google.cloud.location.Locations`
+- `google.iam.v1.IAMPolicy`
+- `google.longrunning.Operations`
+
+**Note:** The list of supported mix-in services and the mix-in services
+themselves change infrequently. Any new common service to be supported should be
+added to the list here and generators should be updated.
+
+To be generated into a client library, a mix-in service **must** be declared
+under [apis] in a [google.api.Service]. Furthermore, it **must** have
+[google.api.http] bindings declared for all of its RPCs in the [http]
+configuration of the same [google.api.Service] (more details on these bindings
+in AIP-127).
+
+
+## Implementing mix-in support
+
+### Generator configuration
+
+Client library generators **must** accept the file path of a
+[google.api.Service] in YAML form. This flag **must** be optional - meaning
+the flag **must not** be required for basic client generation that omits the
+mix-in methods.
+
+### Mix-in API client configuration
+
+If a mix-in API utilizes client library configuration such as
+`google.api.method_signature` or request header injection, generators **may**
+support it in the generated mix-in methods if the configuration is accessible.
+
+### Generating mix-in methods
+
+The mix-in API RPCs **should** be generated as methods on the surface of the
+host API's service client library, alongside the host service's RPCs. This
```

Comment:
That's a good point, a separate client class may be better. Do you mean a single "mixin" client class with all of the configured mixin methods on it? Or do you mean individual `OperationsClient`, `LocationsClient`, `IAMPolicy` client classes? I think the former is better than the latter - it's clear these are utility services, vs. first class product services, and reduces confusion with any existing generated instances of these service clients.

I have a draft implementation of this for Go and while it does generate a bit of duplicate code when there are multiple services, I like that:
- I don't need to instantiate another client to utilize any of these RPCs, they are right there on my existing client 
- I can use the GAPIC as an interface implementation for helpers, like with Operation wrappers that need a client reference to invoke `Poll`, `Wait`, etc. rather than needing to incorporate another client - (ping me if you'd like the internal one-pager I have for some Go GAPIC improvements)
- These mix-in methods are closer to the other helpers, like response-type specific Operation wrappers, Resources a user wants to check the IAM Policy of, or the supported Locations for the Resource they are about to create
- I assert that naming collisions will seldom happen (outside of the known GetIamPolicy, but this the _same_ RPC, not a different one with the same name) 

@miraleung you've implemented this for Java, WDYT?

Reply:
Or, should we be less prescriptive and allow languages to do what they think is best?

Reply:
IMO the current approach has upsides, and I'd like to see what product teams or Yoshi think about the various options (wrt usability). Do we need to consider existing timelines of teams who are asking for this feature?

Backwards API compatibility would be one motivation for adding this to existing clients. In order to migrate PubSub and KMS, they will require either this solution or a one-off among all microgenerators. Moreover, this might allow us to migrate other APIs that copy-pasted IAM RPCs into their protobufs, if a scenario arose where we'd be OK with breaking protobuf changes but none in the generated clients.

For Java and PHP, IMO some upsides of the current approach include a simpler generation logic, client library dev users include not having to figure out which new client to use (for option 1), and they won't be confused between many APIs' LocationsClient (option 2). 

>  _Or, should we be less prescriptive and allow languages to do what they think is best?_

IMO staying consistent with the high-level structure would be ideal, unless there are core language motivations to do otherwise.

Reply:
@lukesneeringer wdyt?

Reply:
> Backwards API compatibility would be one motivation for adding this to existing clients - PubSub and KMS currently use† the existing approach in Java and PHP.

I think we should ignore this point. I am skeptical of the idea that we make sweeping decisions on a design based on 2 existing libraries in 2 languages.

> Moreover, this might allow us to migrate other APIs that copy-pasted IAM RPCs into their protobufs

But _those APIs themselves_ can not migrate for their own reasons, so whether you can migrate them in client libraries is not really relevant.

Reply:
> I am skeptical of the idea that we make sweeping decisions on a design based on 2 existing libraries in 2 languages.

Fair! It is a nice-to-have part of this approach.

> But those APIs themselves can not migrate for their own reasons,

Ok. With a certain API framework being made the new default and precluding the need to redefine such RPCs at the Service level, should we not follow suit with the precedent already set by APIs today that is redefining the RPCs at the Service level? That approach was probably chosen for some technical reasons, but it also brought usability benefits - having the IAM rpcs defined directly on the client.

What say you to the other advantages we believe this approach brings?

Reply:
@noahdietz @miraleung I totally forgot about this thread, and I apologize. I realize I am holding you up.

I suggest let's talk through this in a meeting. I am reasonably convinced in my own mind (😛 ) that my approach is better, but I might be missing something. That said, I think the right decision here is for me to try to convince you, and then defer to you on the final decision.

Seem reasonable?

Reply:
Sounds reasonable to me! I will schedule some time. Thanks Luke.

Reply:
Circling back around after some time now that some blocking tasks have been resolved. I have updated this to allow for language idiomatic representations. Individual language actools+yoshi pairs have indicated a preference that is different from what Mira & I have proposed, and I believe it is reasonable to allow for language idiomatic flexibility - that is the spirit of the micro-generators and a goal of client libraries anyways.